### PR TITLE
Better error messages when accidentally using yield in coroutines

### LINKED ISSUE #1512: Improve error reporting for `yield` inside async function
Those that are in the process of converting `cocotb.coroutine` functions using `yield` to async functions may run into problems where they miss a `yield`, which turns their coroutine into an async generator. If an async generator makes it to the scheduler it will result in a less than helpful error message. Instead we should check for async generators where we expect coroutines and throw a `TypeError` with a more helpful message, much like we do with old style coroutines missing `yield` statements.

### PR DESCRIPTION
**Pull request by ktbarrett**
Fixes #1512. There are a few points where we expect coroutines, so each place got a warning.

### CODE REVIEW

File: tests/test_cases/test_cocotb/test_async_generators.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+
+async def whoops_async_generator():
+    yield cocotb.triggers.Timer(1)
+
+
+@cocotb.test()
+def test_yielding_accidental_async_generator(dut):
+    yield cocotb.triggers.NullTrigger()
+
```
Comment (eric-wieser) [inline tests/test_cases/test_cocotb/test_async_generators.py]: Is this needed?

File: tests/test_cases/test_cocotb/test_async_generators.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+import cocotb
+
+
+async def whoops_async_generator():
+    yield cocotb.triggers.Timer(1)
+
+
+@cocotb.test()
+def test_yielding_accidental_async_generator(dut):
```
Comment (eric-wieser) [inline tests/test_cases/test_cocotb/test_async_generators.py]:
```suggestion
@cocotb.test()
def test_yielding_accidental_async_generator(dut):
    # this test deliberately does not use `async def`, as we are testing the behavior of `yield`
```

File: tests/test_cases/test_cocotb/test_async_generators.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
+import cocotb
+
+
+async def whoops_async_generator():
+    yield cocotb.triggers.Timer(1)
```
Comment (eric-wieser) [inline tests/test_cases/test_cocotb/test_async_generators.py]:
```suggestion
    # the user should have used `await` here, but they wrote `yield` by accident.
    yield cocotb.triggers.Timer(1)
```

File: cocotb/decorators.py [hunk 4]
```diff
@@ -85,12 +85,18 @@ class RunningTask(object):
         task.kill() will destroy a coroutine instance (and cause any Join
         triggers to fire.
     """
+
```
Comment (eric-wieser) [inline cocotb/decorators.py]: Unless this is deliberate:
```suggestion
```

Comment (ktbarrett) [reply cocotb/decorators.py]: It was deliberate. I know it is errant, but my IDE is highlighting the line (and a bunch of others) and it's starting to bother me.

File: tests/test_cases/test_cocotb/test_async_generators.py (hunk 1 above)
Comment (ktbarrett) [reply tests/test_cases/test_cocotb/test_async_generators.py]: Nope.

**Comment by eric-wieser** [comment]
Squashed the commits since the test doesn't pass without the change anyway.

**Pull request merged**